### PR TITLE
Discoveryservice Security

### PR DESCRIPTION
**Pull request by Martin187187**
# Access Control Middleware

This pull request addresses [Issue #45](https://github.com/eclipse-basyx/basyx-go-components/issues/45).  
All **acceptance criteria** have been met.

## Not fully completed
- More **comprehensive security tests** will be added in future iterations.

---

## Summary of changes

This PR introduces a **middleware-based access control system** that enforces the **IDTA-defined access control model**.

### Key features
- ✅ Middleware for endpoint-based access control  
- ✅ Implementation of the **IDTA Access Control Model**  
- ✅ **Keycloak** used as the identity provider  
- ✅ **JWT-based authentication**  
- ✅ **ABAC (Attribute-Based Access Control)** support  
- ✅ Complex access rules supported (based on token claims, time, etc.)  
- ✅ Access rules defined via **JSON** configuration files  
- ✅ Syntax validation for access rule definitions  
- ✅ Rule evaluator capable of processing logical formulas in rules  
- ✅ Initial setup example provided under `/sql_examples/server_secured`

### Current limitations
- Currently supports **Route objects** only (e.g., Discovery Service is too simple for `IdentifiableObject` handling).

---

## How to test

1. Start the secured server example:  
   ```bash
   cd sql_examples/server_secured
   docker compose up -d
   ```
2. Start Discovery Service (aviable in launch)

3. Obtain a token from Keycloak:  
   ```bash
   curl -X POST http://localhost:8080/realms/basyx/protocol/openid-connect/token \
     -H "Content-Type: application/x-www-form-urlencoded" \
     -d "grant_type=password" \
     -d "client_id=basyx-ui" \
     -d "username=martin" \
     -d "password=admin"
   ```

4. Use the retrieved token to test access to secured endpoints.

5. Modify access rules under:  
   ```
   cmd/discoveryservice/config/access_rules/access-rules.json
   ```

5. Adjust user permissions in **Keycloak** to validate different access scenarios.

---

## Notes
- This implementation lays the groundwork for flexible and extensible access control using ABAC principles.
- Additional tests and refinements (especially for security edge cases) will follow.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -5,33 +5,50 @@ go 1.25.0
 require github.com/go-chi/chi/v5 v5.2.3
 
 require (
+	github.com/coreos/go-oidc v2.4.0+incompatible
+	github.com/doug-martin/goqu/v9 v9.19.0
```
Comment (Martin187187) [inline go.mod]: why is goqu import suddenly here?

File: go.mod [hunk 2]
```diff
@@ -5,33 +5,50 @@ go 1.25.0
...
 	github.com/go-chi/cors v1.2.2
 	github.com/jackc/pgx/v5 v5.7.6
 	github.com/lib/pq v1.10.9
 	github.com/spf13/viper v1.21.0
 	github.com/stretchr/testify v1.11.1
+	github.com/xeipuuv/gojsonschema v1.2.0
 )
 
 require (
+	dario.cat/mergo v1.0.2 // indirect
```
Comment (Martin187187) [inline go.mod]: ?

File: cmd/discoveryservice/main.go [hunk 3]
```diff
@@ -105,147 +167,19 @@ func main() {
...
-	// Create a copy of the config to avoid modifying the original
-	cfgCopy := *cfg
-
-	// Redact sensitive information if present in the Postgres configuration
-	if cfg.Postgres.Host != "" {
-		// Simple redaction that preserves the structure but hides credentials
-		cfgCopy.Postgres.Host = "****"
-		cfgCopy.Postgres.User = "****"
-		cfgCopy.Postgres.Password = "****"
+func normalizeBasePath(p string) string {
```
Comment (FriedJannik) [inline cmd/discoveryservice/main.go]: This should be in common

File: internal/common/security/DISCOVERY_AUTHORIZATION_README.md [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+7. **Response Path:** Data returned by the database is further filtered if necessary before being returned to the client.
+
+**Summary:**  
+Even after successful authorization, a user may see only partial data or be limited to certain modifications — thanks to QueryFilters.
+
+---
+
+## 4. ABAC Decision Flow
+
+![ABAC Decision Flow Diagram](./diagrams/abac_flow.svg)
```
Comment (FriedJannik) [inline internal/common/security/DISCOVERY_AUTHORIZATION_README.md]: Maybe put the Mermaid 'Code' inline and don't use pre-generated SVGs - GitHub supports Mermaid (for Version Control)

File: internal/common/security/DISCOVERY_AUTHORIZATION_README.md [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+## 8. Notes
+
+- The ABAC and QueryFilter system ensures **defense in depth**: requests are validated at both the API and data layers.  
+- Access rules can evolve without code changes — only configuration updates are required.  
+- Logging each ABAC decision helps audit access behavior and identify misconfigurations.  
+- QueryFilters are a flexible mechanism to integrate **multi-tenancy**, **privacy filtering**, and **data partitioning** directly into the authorization layer.
+
+---
+
+© 2025 Discovery Service Documentation
```
Comment (FriedJannik) [inline internal/common/security/DISCOVERY_AUTHORIZATION_README.md]:
```suggestion
© 2025 Fraunhofer IESE - Discovery Service Documentation
```

File: internal/common/security/arm.go
Comment (FriedJannik) [inline internal/common/security/arm.go]: Please add GoDoc for Public functions/methods

File: internal/common/security/authorize.go
Comment (FriedJannik) [inline internal/common/security/authorize.go]: Please add GoDoc

File: internal/common/security/oidc.go [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+			return
+		}
+
+		var issuedAt time.Time
+		if n, ok := c["iat"].(json.Number); ok {
+			sec, _ := n.Int64()
+			issuedAt = time.Unix(sec, 0)
+			log.Printf("🕓 Token issued at %v", issuedAt)
+		} else {
+			log.Printf("⚠️ Token missing 'iat' claim")
```
Comment (FriedJannik) [inline internal/common/security/oidc.go]: Should we really ignore it if a token has no iat?

File: sql_examples/server/docker-compose.yml [hunk 7]
```diff
@@ -17,3 +23,19 @@ services:
       - "8080:8080"
     depends_on:
       - db
+  keycloak:
```
Comment (FriedJannik) [inline sql_examples/server/docker-compose.yml]: Why do we need keycloak here - isn't the docker_compose in server_secured sufficient enough?

File: sql_examples/server/docker-compose.yml [hunk 8]
```diff
@@ -9,6 +9,12 @@ services:
     ports:
       - "5432:5432"
     command: ["postgres", "-c", "listen_addresses=*"]
+    
+    healthcheck:
```
Comment (FriedJannik) [inline sql_examples/server/docker-compose.yml]: See comment regarding keycloak

**Review by FriedJannik — CHANGES_REQUESTED** [review]

**Review by FriedJannik — APPROVED** [review]
marcus-the-worm
"_Robert, i approve this Pull Request_"
[2025-10-28 16-44-41.mp3](https://github.com/user-attachments/files/23191939/2025-10-28.16-44-41.mp3)
<video controls="" autoplay="" name="media">
    <source src="https://github.com/user-attachments/files/23191939/2025-10-28.16-44-41.mp3" type="audio/mp3">
</video>
